### PR TITLE
Replace NSURLConnection for NSURLSession

### PR DESCRIPTION
**Pull request by rhyshort**
## What 

Replace `NSURLConnection` with `NSURLSession`. Also Includes initial filtering and context classes.

## Why

`NSURLSession` gives us more flexibility, we no longer have to use the shared cookie store for the application and we can separate out our requests from any HTTP requests they make to their backend.

## How

- `TDRemoteRequest` (and its' subclasses ) moved to using `CDTURLSession`, a facade for `NSURLConnection`.
- `TDURLConnectionChangeTracker` (Probably needs remaining now) moved to `NSURLSession`
- Bump the supported platforms from iOS 6 to iOS 7 and OS X 10.8 to 10.9

 
## Limitations 
- Completion handler callbacks occur on the thread that created the `CDTURLSession` object. This is a temporary stop gap until the `TDReplicator` (and it's sub classes) can be made thread safe.
- Filters are not completely implemented, the classes are there as part of the initial import for @tomblench's initial test project.

## Reviewers

reviewer @tomblench 
reviewer @gadamc 
reviewer @mikerhodes

### CODE REVIEW

File: Classes/common/CDTDatastore.m [hunk 1]
```diff
@@ -324,6 +324,12 @@ - (NSArray *)allDocsQuery:(NSArray *)docIds options:(TDQueryOptions *)queryOptio
     for (NSDictionary *row in dictResults[@"rows"]) {
         //            NSLog(@"%@", row);
         NSString *docId = row[@"id"];
+        
+        if(!docId){
+            NSLog(@"Document not found info: %@",row);
```
Comment (tomblench) [inline Classes/common/CDTDatastore.m]: debugging code left in?

File: Classes/common/touchdb/TDRemoteRequest.m [hunk 2]
```diff
@@ -174,80 +183,22 @@ - (BOOL)retry
     // request might have succeeded on the remote server, and by retrying we'd be issuing it again.
     // PUT and POST requests aren't generally idempotent, but the ones sent by the replicator are.
 
-    if (_retryCount >= kMaxRetries) return NO;
-    NSTimeInterval delay = RetryDelay(_retryCount);
-    ++_retryCount;
-    CDTLogVerbose(CDTTD_REMOTE_REQUEST_CONTEXT, @"%@: Will retry in %g sec", self, delay);
-    [self startAfterDelay:delay];
-    return YES;
+//    if (_retryCount >= kMaxRetries) return NO;
```
Comment (tomblench) [inline Classes/common/touchdb/TDRemoteRequest.m]: There's a load of deleted / commented out code in here around retries and authentication. I think we need to figure out if it's a bad thing to lose this functionality and what functionality can be done by the filterchain and what can be done by NSURLSession

File: Classes/common/touchdb/TDMultipartDownloader.m [hunk 3]
```diff
@@ -70,18 +65,15 @@ - (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLRespon
...
-    [super connection:connection didReceiveResponse:response];
+    
+    [super receivedResponse:response];
 }
 
-- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data
+- (void)receivedData:(NSData *)data
 {
-    [super connection:connection didReceiveData:data];
+   // [super connection:connection didReceiveData:data];
```
Comment (tomblench) [inline Classes/common/touchdb/TDMultipartDownloader.m]: call [super ] ?

File: Classes/common/touchdb/TDMultipartUploader.m [hunk 4]
```diff
@@ -67,7 +67,7 @@ - (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)err
         NSError *writerError = _multipartWriter.error;
         if (writerError) error = writerError;
     }
-    [super connection:connection didFailWithError:error];
+    //[super connection:connection didFailWithError:error];
```
Comment (tomblench) [inline Classes/common/touchdb/TDMultipartUploader.m]: call [super ]?

Comment (rhyshort) [inline Classes/common/touchdb/TDMultipartUploader.m]: done in 6d15b5a

File: Classes/common/touchdb/TDMultipartDownloader.m (hunk 3 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDMultipartDownloader.m]: done in a3cb86a

File: Classes/common/CDTDatastore.m (hunk 1 above)
Comment (rhyshort) [inline Classes/common/CDTDatastore.m]: removed in 1d8cb9a

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 2 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: Retires should be handled by `CDTURLSession` I haven't figured out how though. The authentication methods changed, most of it is still present, but for accepting SSL certs currently only the default NSURLSession behaviour is observed. We should have it back to feature parity with the previous implementation but I wonder if there was a way to configure the SSL cert verification anyway

File: ReplicationAcceptance/ReplicationAcceptance.xcodeproj/xcshareddata/xcschemes/RA_Tests_OSX.xcscheme [hunk 5]
```diff
@@ -75,6 +75,46 @@
...
+         <AdditionalOption
+            key = "MallocGuardEdges"
+            value = ""
+            isEnabled = "YES">
+         </AdditionalOption>
+         <AdditionalOption
+            key = "MallocScribble"
+            value = ""
+            isEnabled = "YES">
+         </AdditionalOption>
```
Comment (mikerhodes) [inline ReplicationAcceptance/ReplicationAcceptance.xcodeproj/xcshareddata/xcschemes/RA_Tests_OSX.xcscheme]: I don't think these debugging options should have been committed.

File: ReplicationAcceptance/ReplicationAcceptance.xcodeproj/xcshareddata/xcschemes/RA_Tests.xcscheme [hunk 6]
```diff
@@ -66,6 +66,46 @@
...
+         <AdditionalOption
+            key = "MallocGuardEdges"
+            value = ""
+            isEnabled = "YES">
+         </AdditionalOption>
+         <AdditionalOption
+            key = "MallocScribble"
+            value = ""
+            isEnabled = "YES">
+         </AdditionalOption>
```
Comment (mikerhodes) [inline ReplicationAcceptance/ReplicationAcceptance.xcodeproj/xcshareddata/xcschemes/RA_Tests.xcscheme]: I don't think these debugging options should have been committed.

File: ReplicationAcceptance/ReplicationAcceptance/ReplicationAcceptance.m [hunk 7]
```diff
@@ -1636,59 +1642,60 @@ -(void) testURLConnectionChangeTrackerWithBadCredentials
...
+//                                 beforeDate: [NSDate dateWithTimeIntervalSinceNow:0.1]];
+//    }
+//    
+//    [NSURLProtocol unregisterClass:[ChangeTrackerNSURLProtocolTimedOut class]];
+//    
+//    XCTAssertFalse(changeTrackerGotChanges);
+//    NSUInteger retryCount = [(TDURLConnectionChangeTracker *)changeTracker totalRetries];
+//    XCTAssertTrue(retryCount == 6, @"Expected kMaxRetries(6) retries, found %ld",
+//                  (unsigned long)retryCount);
+//}
```
Comment (mikerhodes) [inline ReplicationAcceptance/ReplicationAcceptance/ReplicationAcceptance.m]: Removal of a test should be in its own commit with an explanation as to why you removed it.

File: ReplicationAcceptance/ReplicationAcceptance.xcodeproj/xcshareddata/xcschemes/RA_Tests_OSX.xcscheme (hunk 5 above)
Comment (rhyshort) [inline ReplicationAcceptance/ReplicationAcceptance.xcodeproj/xcshareddata/xcschemes/RA_Tests_OSX.xcscheme]: fixed in bf89ac6

File: ReplicationAcceptance/ReplicationAcceptance.xcodeproj/xcshareddata/xcschemes/RA_Tests.xcscheme (hunk 6 above)
Comment (rhyshort) [inline ReplicationAcceptance/ReplicationAcceptance.xcodeproj/xcshareddata/xcschemes/RA_Tests.xcscheme]: fixed in bf89ac6

File: ReplicationAcceptance/ReplicationAcceptance/ReplicationAcceptance.m (hunk 7 above)
Comment (rhyshort) [inline ReplicationAcceptance/ReplicationAcceptance/ReplicationAcceptance.m]: done, now in commit 9292f55

File: Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m [hunk 8]
```diff
@@ -81,17 +86,18 @@ - (BOOL)start
...
+    self.task = [self.session dataTaskWithRequest:self.request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+        if(error){
+            [self failedWithError:error];
+            return;
+        } else {
+            [self receivedResponse:response];
+            [self receivedData:data];
+            [self finishedLoading];
+        }
+    }];
```
Comment (mikerhodes) [inline Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m]: Is this use of `self` in the block safe?

File: Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m [hunk 9]
```diff
@@ -103,8 +109,10 @@ - (BOOL)start
 
 - (void)clearConnection
 {
-    [self.connection cancel];
-    [self.connection unscheduleFromRunLoop:[NSRunLoop currentRunLoop] forMode:NSRunLoopCommonModes];
+    if(self.task.state != NSURLSessionTaskStateCompleted){
+        [self.task cancel];
+    }
+    self.task = nil;
     self.connection = nil;
```
Comment (mikerhodes) [inline Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m]: Why is this line still here if the other connection things are gone? I also don't see the removal of the connection property.

File: Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m [hunk 10]
```diff
@@ -172,7 +178,8 @@ - (void)connection:(NSURLConnection *)connection willSendRequestForAuthenticatio
         
         CDTLogVerbose(CDTREPLICATION_LOG_CONTEXT, @"%@ challenge: continueWithoutCredential",
                       [self class]);
-        [sender continueWithoutCredentialForAuthenticationChallenge:challenge];
+        completionHandler(NSURLSessionAuthChallengeUseCredential,nil);
+//        [sender continueWithoutCredentialForAuthenticationChallenge:challenge];
```
Comment (mikerhodes) [inline Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m]: Are these lines equivalent, ditto for all the below ones.

File: Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m [hunk 11]
```diff
@@ -273,14 +286,18 @@ -(void) connectionDidFinishLoading:(NSURLConnection *)connection
     
     [self clearConnection];
     
-    if (restart)
+    if (restart){
+        NSLog(@"Polling _changes again");
```
Comment (mikerhodes) [inline Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m]: Remove NSLog statement.

File: Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m [hunk 12]
```diff
@@ -273,14 +286,18 @@ -(void) connectionDidFinishLoading:(NSURLConnection *)connection
...
         [self start];  // Next poll...
-    else
+    }
+    else {
+        NSLog(@"Stopping chnage tracker");
```
Comment (mikerhodes) [inline Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m]: Remove NSLog statement

File: Classes/common/touchdb/TDMultipartDownloader.m [hunk 13]
```diff
@@ -37,25 +37,20 @@ - (id)initWithURL:(NSURL *)url
                     onCompletion:onCompletion];
     if (self) {
         _db = database;
+        _reader = [[TDMultipartDocumentReader alloc] initWithDatabase:_db];
+        [_request setValue:@"multipart/related, application/json" forHTTPHeaderField:@"Accept"];
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDMultipartDownloader.m]: The `setUpRequest:withBody:` method assigned a body, need we to do that here?

File: Classes/common/touchdb/TDRemoteRequest.h [hunk 14]
```diff
@@ -57,10 +52,12 @@ typedef void (^TDRemoteRequestCompletionBlock)(id result, NSError* error);
 @property (readonly) NSMutableDictionary* statusInfo;
 
 // protected:
-- (void)setupRequest:(NSMutableURLRequest*)request withBody:(id)body;
 - (void)clearConnection;
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.h]: Change the name, there isn't a connection any more.

File: Classes/common/touchdb/TDRemoteRequest.m [hunk 15]
```diff
@@ -31,11 +31,20 @@
...
+#import "CDTURLSession.h"
 
 // Max number of retry attempts for a transient failure, and the backoff time formula
 #define kMaxRetries 2
 #define RetryDelay(COUNT) (4 << (COUNT))  // COUNT starts at 0
 
+@interface TDRemoteRequest()
+
+@property CDTURLSession * session;
+@property NSURLSessionDataTask * task;
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: nonatomic, strong

Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: spacing

File: Classes/common/touchdb/TDRemoteRequest.m [hunk 16]
```diff
@@ -48,7 +57,7 @@ + (NSString *)userAgentHeader
 #elif TARGET_OS_MAC
     platform = @"Mac OS X";
 #endif
-    return $sprintf(@"CloudantSync/%@ (%@ %@)", [CDTDatastore versionString], platform, version);
+    return [NSString stringWithFormat:@"CloudantSync/%@ (%@ %@)", [CDTDatastore versionString], platform, version ];
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: wrapping

File: Classes/common/touchdb/TDRemoteRequest.m [hunk 17]
```diff
@@ -65,12 +74,11 @@ - (id)initWithMethod:(NSString *)method
         _request.cachePolicy = NSURLRequestReloadIgnoringLocalCacheData;
 
         // Add headers.
-        [_request setValue:[[self class] userAgentHeader] forHTTPHeaderField:@"User-Agent"];
+        [_request setValue:[TDRemoteRequest userAgentHeader] forHTTPHeaderField:@"User-Agent"];
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: `self class` to allow subclasses to override.

File: Classes/common/touchdb/TDRemoteRequest.m [hunk 18]
```diff
@@ -65,12 +74,11 @@ - (id)initWithMethod:(NSString *)method
...
         [requestHeaders enumerateKeysAndObjectsUsingBlock:^(id key, id value, BOOL *stop) {
             [_request setValue:value forHTTPHeaderField:key];
         }];
-
-        [self setupRequest:_request withBody:body];
+        _session = [[CDTURLSession alloc]init];
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: spacing

File: Classes/common/touchdb/TDRemoteRequest.m [hunk 19]
```diff
@@ -86,72 +94,73 @@ - (void)setAuthorizer:(id<TDAuthorizer>)authorizer
...
-// Retaining myself shouldn't be necessary, because NSURLConnection is documented as retaining
-// its delegate while it's running. But GNUstep doesn't (currently) do this, so for
-// compatibility I retain myself until the connection completes (see -clearConnection.)
-// TODO: Remove this and the [self autorelease] below when I get the fix from GNUstep.
-#ifdef GNUSTEP
-    [self retain];
-#endif
+
+    __weak TDRemoteRequest * weakSelf = self;
+    self.task = [self.session dataTaskWithRequest:_request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: spacing above, line length this line

File: Classes/common/touchdb/TDRemoteRequest.m [hunk 20]
```diff
@@ -86,72 +94,73 @@ - (void)setAuthorizer:(id<TDAuthorizer>)authorizer
...
+        // the response has actually completed.
+        TDRemoteRequest * strongSelf = weakSelf;
+        if(error){
+            [strongSelf requestDidError:error];
+        } else if (TDStatusIsError(((NSHTTPURLResponse*)response).statusCode)) {
+                        [strongSelf receivedResponse:response];
+        }  else {
+            [strongSelf receivedResponse:response];
+            [strongSelf receivedData:data];
+        }
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: spacing throughout the if

File: Classes/common/touchdb/TDRemoteRequest.m [hunk 21]
```diff
@@ -86,72 +94,73 @@ - (void)setAuthorizer:(id<TDAuthorizer>)authorizer
...
-// TODO: Remove this and the [self autorelease] below when I get the fix from GNUstep.
-#ifdef GNUSTEP
-    [self retain];
-#endif
+
+    __weak TDRemoteRequest * weakSelf = self;
+    self.task = [self.session dataTaskWithRequest:_request completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
+        //we need to fire all the delegate methods kinda of....
+        //we actually need to replace the delegate methods with resonable things to do since
+        // the response has actually completed.
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: Comment nonsensical

File: Classes/common/touchdb/TDRemoteRequest.m [hunk 22]
```diff
@@ -86,72 +94,73 @@ - (void)setAuthorizer:(id<TDAuthorizer>)authorizer
...
-    return $mdict({ @"URL", _request.URL.absoluteString }, { @"method", _request.HTTPMethod });
+    return [@{ @"URL": _request.URL.absoluteString , @"method": _request.HTTPMethod } mutableCopy];
 }
 
 - (void)respondWithResult:(id)result error:(NSError *)error
 {
     Assert(result || error);
+    if(_onCompletion){
     _onCompletion(result, error);
+    }
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: spacing

File: Classes/common/touchdb/TDRemoteRequest.m [hunk 23]
```diff
@@ -86,72 +94,73 @@ - (void)setAuthorizer:(id<TDAuthorizer>)authorizer
...
     _onCompletion = nil;  // break cycles
 }
 
 - (void)startAfterDelay:(NSTimeInterval)delay
 {
     // assumes _connection already failed or canceled.
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: Fix comments

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 2 above)
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: This feels like it was commented out just because it didn't work. Why doesn't it work?

File: Classes/common/touchdb/TDRemoteRequest.m [hunk 24]
```diff
@@ -174,80 +183,22 @@ - (BOOL)retry
...
-        } else {
-            CDTLogVerbose(CDTTD_REMOTE_REQUEST_CONTEXT, @"    challenge: cancel");
-            [sender cancelAuthenticationChallenge:challenge];
-        }
-    } else {
-        CDTLogVerbose(CDTTD_REMOTE_REQUEST_CONTEXT, @"    challenge: performDefaultHandling");
-        [sender performDefaultHandlingForAuthenticationChallenge:challenge];
-    }
-}
-
```
Comment (mikerhodes) [inline Classes/common/touchdb/TDRemoteRequest.m]: I don't think we can just remove this?

**Comment by mikerhodes** [comment]
As @tomblench says, this PR removes a bunch of functionality without explaining why -- which would allow for deciding whether it's the correct thing to do at this point. Without a rationale for this, it's not possible to merge the PR because it's doing stuff I don't understand.

File: Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m (hunk 9 above)
Comment (rhyshort) [inline Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m]: 966b52e

File: Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m (hunk 10 above)
Comment (rhyshort) [inline Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m]: I believe so yes. removed commented out lines in d457d22

File: Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m (hunk 11 above)
Comment (rhyshort) [inline Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m]: done in b5cf9b1

File: Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m (hunk 12 above)
Comment (rhyshort) [inline Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m]: done in b5cf9b1

File: Classes/common/touchdb/TDMultipartDownloader.m (hunk 13 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDMultipartDownloader.m]: No, multipart downloader should only be used with GET requests irrc

File: Classes/common/touchdb/TDRemoteRequest.h (hunk 14 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.h]: done in ab46ec2

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 15 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: nonatomic done in b65a640 and spacing in 42e40dd

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 16 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: done in d544a44

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 17 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: done in d8d6e05

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 18 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: done in 1bc4b7c

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 19 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: done in c229418

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 20 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: done in 7bedbb2

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 21 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: removed in 41dfb8d

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 22 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: done in 2c6b483

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 23 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: fixed comment on line 154 in b62e2e4

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 2 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: It should work, but I commented it out rather than delete it to make it easier to go back if I broke something, and tbh with all the changes to try and get this working I forgot I had it commented out.

File: Classes/common/touchdb/TDRemoteRequest.m (hunk 24 above)
Comment (rhyshort) [inline Classes/common/touchdb/TDRemoteRequest.m]: added back, using the correct APIs for NSURLSession in 296e7ce

File: Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m (hunk 8 above)
Comment (rhyshort) [inline Classes/common/touchdb/ChangeTracker/TDURLConnectionChangeTracker.m]: I'm not sure, I've converted it to use the normal `self` to `weakSelf` to `strongSelf` pattern in d1e9e05

File: Classes/common/HTTP/CDTURLSession.m [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+}
+
+-(instancetype)initWithDelegate:(id<NSURLSessionDelegate>)delegate{
+    self = [super init];
+    if (self) {
+        _thread = [NSThread currentThread];
+        
+        NSURLSessionConfiguration *config = [NSURLSessionConfiguration defaultSessionConfiguration];
+        
+        _session = [NSURLSession sessionWithConfiguration:config delegate:delegate delegateQueue:[NSOperationQueue currentQueue]];
```
Comment (mikerhodes) [inline Classes/common/HTTP/CDTURLSession.m]: Why not create a serial queue here and avoid the MYOnThread stuff -- does that not work?

**Comment by rhyshort** [comment]
Opening new pr on feature branch
